pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,275 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains functions for preprocessing `QuantumScript`s to ensure
+that they are supported for execution by a device."""
+
+import pennylane as qml
+
+# from pennylane.tape import QuantumScript
+from pennylane.operation import Observable, Tensor
+from pennylane.measurements import (
+    MidMeasureMP,
+    ExpectationMP,
+    ShadowExpvalMP,
+    SampleMP,
+    ProbabilityMP,
+    CountsMP,
+)
+from pennylane import DeviceError
+
+
+def _stopping_condition(op):
+    """Specify whether or not an Operator object is supported by the device"""
+    return getattr(op, "has_matrix", False) and qml.matrix(op).shape[-1] % 2 == 0
+
+
+def _supports_observable(dev, observable):
+    """Checks if an observable is supported by this device. Raises a ValueError,
+        if not a subclass or string of an Observable was passed.
+
+    Args:
+        dev (.Device): device to check
+        observable (type or str): observable to be checked
+
+    Returns:
+        bool: ``True`` iff supplied observable is supported
+    """
+    if isinstance(observable, type) and issubclass(observable, Observable):
+        observable = observable.__name__
+    if isinstance(observable, str):
+        return observable in dev.observables
+
+    return False
+
+
+def expand_fn(circuit, dev, max_expansion=10):
+    """Method for expanding or decomposing an input circuit. Can be the default or
+    a custom expansion method, see :meth:`.Device.custom_expand` for more details.
+
+    By default, this method expands the tape if:
+
+    - nested tapes are present,
+    - any operations are not supported on the device, or
+    - multiple observables are measured on the same wire.
+
+    Args:
+        circuit (.QuantumTape): the circuit to expand.
+        dev (.Device): the device to execute circuit(s) on.
+        max_expansion (int): The number of times the circuit should be
+            expanded. Expansion occurs when an operation or measurement is not
+            supported, and results in a gate decomposition. If any operations
+            in the decomposition remain unsupported by the device, another
+            expansion occurs.
+
+    Returns:
+        .QuantumTape: The expanded/decomposed circuit, such that the device
+        will natively support all operations.
+    """
+    # pylint: disable=protected-access
+
+    if dev.custom_expand_fn is not None:
+        return dev.custom_expand_fn(circuit, max_expansion=max_expansion)
+
+    comp_basis_sampled_multi_measure = (
+        len(circuit.measurements) > 1 and circuit.samples_computational_basis
+    )
+    obs_on_same_wire = len(circuit._obs_sharing_wires) > 0 or comp_basis_sampled_multi_measure
+    obs_on_same_wire &= not any(isinstance(o, qml.Hamiltonian) for o in circuit._obs_sharing_wires)
```

Comment:
So if you look at #3700 , you can see we will be supporting simultaneous simulation of non-commuting measurements.  For example, we will be able to simulate
```
qml.tape.QuantumScript(measurements=[qml.expval(qml.PauliZ(0)), qml.PauliY(0))])
```
in one pass.

So "observables on the same wire" is no longer a concern for the new Python device.

Reply:
In the prototype device in #3457, I noticed that the measurements only support `StateMeasurement` types. Should I also remove the sample-based logic from the code in this PR?